### PR TITLE
virtio_serial:Hotplug chardev and port with a virtconsole

### PR DESCRIPTION
**Pull request by nanliu-r**
ID: 1755676

Signed-off-by: Nana Liu <nanliu@redhat.com>

### CODE REVIEW

**Comment by nanliu-r** [comment]
Test Results:
 (1/1) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci.with_console: PASS (64.07 s)
 (1/1) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci.with_console.q35: PASS (83.55 s)

**Comment by nanliu-r** [comment]
@vivianQizhu Please help to review this, thanks a lot.

File: qemu/tests/hotplug_port_chardev_pci_with_console.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Hot-plug chardev and virtserialport
+
+    1. Boot a guest with 1 virtconsole attached to pty backend
+    2. Hot plug unix_socket backend
+    3. Hot add virtserialport, attached on unix chardev
+    4. Transfer data between host and guest via virtserialport
+    5. Hot-unplug existed virt-serial-device, this should fail
```
Comment (vivianQizhu) [inline qemu/tests/hotplug_port_chardev_pci_with_console.py]: What does this mean? `virtio-serial-pci`? Should fail?

File: qemu/tests/virtio_serial_hotplug_port_pci.py [hunk 2]
```diff
@@ -13,6 +13,21 @@
 from qemu.tests.vioser_in_use import live_migration_guest  # pylint: disable=W0611
 
 
+def get_buses_serial_devices(vm, params, char_devices, serials):
```
Comment (vivianQizhu) [inline qemu/tests/virtio_serial_hotplug_port_pci.py]: Doc string, and recommend to add `and` between `buses` and `serial_devices`

File: qemu/tests/hotplug_port_chardev_pci_with_console.py (hunk 1 above)
Comment (nanliu-r) [reply qemu/tests/hotplug_port_chardev_pci_with_console.py]: Forget to correct it, thanks and update.

File: qemu/tests/virtio_serial_hotplug_port_pci.py (hunk 2 above)
Comment (nanliu-r) [reply qemu/tests/virtio_serial_hotplug_port_pci.py]: Thanks and update.

**Comment by vivianQizhu** [comment]
@nanliu-r Please fix the conflict, thanks.

File: qemu/tests/hotplug_port_chardev_pci_with_console.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+from qemu.tests.virtio_console import add_chardev
+from qemu.tests.virtio_console import add_virtio_ports_to_vm
+from qemu.tests.virtio_serial_file_transfer import transfer_data
+from qemu.tests.virtio_serial_hotplug_port_pci import get_buses_and_serial_devices
+
+
+def get_virtio_serial_pci(vm, serial_device):
+    serial_device_bus = serial_device.get_param('bus')
+    local_index = serial_device_bus.index('.')
+    serial_bus_id = serial_device_bus[:(local_index)]
```
Comment (vivianQizhu) [inline qemu/tests/hotplug_port_chardev_pci_with_console.py]: `serial_bus_id = serial_device_bus.split('.')[0]`

File: qemu/tests/cfg/virtio_port_hotplug.cfg [hunk 4]
```diff
@@ -62,6 +62,15 @@
                     interrupt_test_after_unplug = live_migration_guest
                 - repeat_pci_in_loop:
                     repeat_times = 100
+                - with_console:
```
Comment (vivianQizhu) [inline qemu/tests/cfg/virtio_port_hotplug.cfg]: Recommend name it with `negative`

File: qemu/tests/hotplug_port_chardev_pci_with_console.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    5. Hot-unplug existed virt-serial-pci, this should success without crash
+
+    :param test: kvm test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+    vm = env.get_vm(params['main_vm'])
+    char_devices = add_chardev(vm, params)
+    serials = params.objects('extra_serials')
+    buses, serial_devices = get_buses_and_serial_devices(
```
Comment (vivianQizhu) [inline qemu/tests/hotplug_port_chardev_pci_with_console.py]: `serial_devices = get_buses_and_serial_devices()[1]` if buses is not usable here.

File: qemu/tests/hotplug_port_chardev_pci_with_console.py (hunk 3 above)
Comment (nanliu-r) [reply qemu/tests/hotplug_port_chardev_pci_with_console.py]: thanks and update and will update test result later.

File: qemu/tests/cfg/virtio_port_hotplug.cfg (hunk 4 above)
Comment (nanliu-r) [reply qemu/tests/cfg/virtio_port_hotplug.cfg]: Thanks and update.

File: qemu/tests/hotplug_port_chardev_pci_with_console.py (hunk 5 above)
Comment (nanliu-r) [reply qemu/tests/hotplug_port_chardev_pci_with_console.py]: Thanks and update, will add test result later.

**Comment by nanliu-r** [comment]
> @nanliu-r Please rebase your branch properly.
update and test pass, please help to review again, thanks.

**Comment by nanliu-r** [comment]
Test results:

 (1/1) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci.with_console_negative.q35: PASS (81.20 s)
 (1/1) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.virtio_port_hotplug.hotplug_port_pci.with_console_negative: PASS (73.10 s)

**Pull request merged**